pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,436 @@
+
+    A contributor is ranked by how large of a proportion of all work they did - 
+    if someone authored 350 of 1000 commits, for instance, they'll account for 35% 
+    of the commits made. The list of percent contributions is ranked in descending order. 
+    The ordered list is greedily processed such that the smallest list of contributors is 
+    found that accounts for threshold percent of the work done for a given contribution type.
+   
+    returns the contributor importance, cntrb_id's and number of outliers for the given contribution type
+    '''
+    # identify and count outliers which we define as 2 standard deviations or more above from the mean
+    num_outliers = len(df[(stats.zscore(df["Action"]) > 2)])
+    
+    # calculate total number of contributions
+    t_cntrbs = df["Action"].sum()
+
+    # initilize empty list of to store contibutor ids for those who make up the contributor importance
+    cntrb_ids = []
+    # initialize cumulative percent of contributions
+    cum_per = 0
+
+    # add column name cntrbs to calculate the ratio of activity to the total amount of said activity
+    df["cntrbs"] = df["Action"] / t_cntrbs
+    df = df.reset_index()
+    
+    # iterate through each row in df and create a running sum of percent contributions
+    for _, row in df.iterrows():
+        cum_per += row['cntrbs']
+        cntrb_ids.append(row['cntrb_id']) # add contributor id to list 
+        # if the cumulative percent of contributions is greater than or equal to the threshold, break
+        if cum_per  >= threshold:
+            break
+    
+    # calculate the contributor importance
+    cntrb_importance = len(cntrb_ids)
+    return cntrb_importance, cntrb_ids, num_outliers
+
+def calc_cntrb_importance_over_time(df: pd.DataFrame, patterns, threshold,  window_width, step_size, start_date, end_date):
+      # if the start date and end date are null, set them to the first and last commit date
+    print(df.columns)
+    if start_date == None:
+        start_date = min(df['created_at'])
+       
+    if end_date == None:
+        end_date = max(df['created_at'])
+
+    print(start_date, end_date)
+    # convert the difference between the end date and start date into months
+    delta_year = (end_date.year - start_date.year) 
+    delta_month = (delta_year * 12) + (end_date.month - start_date.month)
+    
+    # if the window width is greater than time spanned by start date and end date, raise ValueError
+    if window_width > delta_month:
+        raise ValueError('The window width cannot be greater than the time spanned by the start and end date.')
+    
+    # if the step size is greater than window width, raise ValueError
+    if step_size > window_width:
+        raise ValueError('The step size cannot be greater than the window width.')
+     
+    window_width=relativedelta(months=window_width)
+    step_size=relativedelta(months=step_size)
```

Comment:
This all can be GREATLY simplified. Look at the process_data function for the active_drifting_contributor or staleness graphs on the overview page. Use pd.date_range to get the "start date"/"period from" for each step and then make a df (just like lines 209 to 216 in issue_staleness) with that as a column, then you can make another column for the end date_period to by just adding the window size to the start date. From there just use a .apply in a very similar fashion as those visualizations